File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Turkey.m
//
//  UIImage+Turkey.m
//  CountriesFlags
//

#import "Countries/UIImage+Turkey.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/UIBezierPath+AffineTransform.h"

@implementation UIImage (Turkey)

+ (UIImage * _Nullable)flagTurkeyWithSize:(CGSize)size {
    UIColor* colorRed = [UIColor colorWithRed:227.0f/255.0f
                                        green:10.0f/255.0f
                                         blue:23.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorWhite setFill];
    [[UIBezierPath bezierPathWithArcCenter:CGPointMake(size.height / 2.0f, size.height / 2.0f)
                                    radius:size.height / 4.0f
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];
    
    [colorRed setFill];
    [[UIBezierPath bezierPathWithArcCenter:CGPointMake(size.height / 2.0f + size.height / 16.0f, size.height / 2.0f)
                                    radius:size.height * 2.0 / 10.0
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0f
                                 clockwise:YES] fill];
    
    [colorWhite setFill];
    CGPoint starCenter = CGPointMake(size.height / 1280.0f * 1050.0f, size.height / 2.0f);
    UIBezierPath* path = [UIBezierPath starAtPoint:starCenter
                                            radius:size.height / 8.0f
                                              type:FlagStarTypeFivePointed];
    [path rotateToAngle:-90.0f center:starCenter];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Turkey.swift
//
//  UIImage+Turkey.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Turkey "wikipedia"
     
     Generates the flag of Turkey.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Turkey.
     */
    internal static func flagTurkey(with size: CGSize) -> UIImage? {
        let colorRed = UIColor(red: 227.0 / 255.0,
                               green: 10.0 / 255.0,
                               blue: 23.0 / 255.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorWhite.setFill()
        UIBezierPath(arcCenter: CGPoint(x: size.height / 2.0, y: size.height / 2.0),
                     radius: size.height / 4.0,
                     startAngle: 0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorRed.setFill()
        UIBezierPath(arcCenter: CGPoint(x: size.height / 2.0 + size.height / 16.0, y: size.height / 2.0),
                     radius: size.height * 2.0 / 10.0,
                     startAngle: 0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        colorWhite.setFill()
        let starCenter = CGPoint(x: size.height / 1280.0 * 1050.0, y: size.height / 2.0)
        let path = UIBezierPath.star(at: starCenter, radius: size.height / 8.0)
        path.rotate(to: -90.0, around: starCenter)
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
